src/synth.rs

```diff
@@ -0,0 +1,422 @@
+            runner = runner.with_expr(&L::instantiate(&eq.rhs));
+        }
+
+        // 3. run eqsat with self.prior_rws and self.new_rws
+        let rewrites = self
+            .prior_rws
+            .values()
+            .map(|eq| &eq.rewrite)
+            .chain(self.new_rws.values().map(|eq| &eq.rewrite));
+        runner = runner.run(rewrites);
+
+        // 4. go through candidates (root pairs) and if they haven't yet
+        // merged, they are still candidates
+        let mut remaining_candidates = EqualityMap::default();
+        let extract = Extractor::new(&runner.egraph, AstSize);
+        for ids in runner.roots.chunks(2) {
+            if runner.egraph.find(ids[0]) == runner.egraph.find(ids[1]) {
+                // candidate has merged (derivable from other rewrites)
+                continue;
+            }
+            let (_, left) = extract.find_best(ids[0]);
+            let (_, right) = extract.find_best(ids[1]);
+            if let Some(eq) = Equality::new(&left, &right) {
+                remaining_candidates.insert(eq.name.clone(), eq);
+            }
+        }
+
+        remaining_candidates
+    }
+
+    fn choose_eqs(&mut self, candidates: EqualityMap<L>) {
+        let step_size = 1;
+        let mut remaining_candidates = candidates;
+        while !remaining_candidates.is_empty() {
+            remaining_candidates = self.select(step_size, remaining_candidates);
+
+            remaining_candidates = self.shrink(&remaining_candidates);
+        }
+    }
+
+    fn run_cvec_synth(&mut self) -> EqualityMap<L> {
+        let eqs = self.prior_rws.clone();
+
+        let runner = self.mk_runner(self.egraph.clone());
+        let (_, unions, _) =
+            self.run_rewrites(runner, eqs.values().map(|eq| &eq.rewrite).collect());
+
+        self.apply_unions(unions);
+
+        self.cvec_match()
+    }
+
+    fn extract_candidates_from_unions(&mut self, unions: HashMap<Id, Vec<Id>>) -> EqualityMap<L> {
+        let mut candidates: EqualityMap<L> = EqualityMap::default();
+        let clone = self.egraph.clone();
+        let extract = Extractor::new(&clone, AstSize); // TODO: cost function for allowed
+        for ids in unions.values() {
+            for id1 in ids.clone() {
+                for id2 in ids.clone() {
+                    let (c1, e1) = extract.find_best(id1);
+                    let (c2, e2) = extract.find_best(id2);
+                    if c1 == usize::MAX || c2 == usize::MAX {
+                        continue;
+                    }
+                    if let Some(eq) = Equality::new(&e1, &e2) {
+                        if e1 != e2 {
+                            // TODO: we used to validate here- do we need to?
+                            candidates.insert(eq.name.clone(), eq);
+                        }
+                    }
+                }
+            }
+        }
+        candidates
+    }
+
+    fn run_rule_lifting(&mut self) -> EqualityMap<L> {
```

Comment:
so we should make sure this matches what we discussed, or maybe we need to go over this again and compare this version against the ones we drafted in the paper.

Reply:
This matches the version on main currently, which is the new version from the meeting notes a few weeks ago